### PR TITLE
enforces translation of GUI strings

### LINKED ISSUE #1265: review codebase for UI strings to be translated
- [x] Look for strings missing `gettext()` wrappers
- [x] Look for string constants that can be factored out
- [x] Keep an eye out for strings that may be fragmented (to accommodate Qt widgets) that may make idiomatic translation difficult
    - [x] Can we add gettext translator comments to supply context to translators working in Weblate?

### PR DESCRIPTION
**Pull request by cfm**
Ready for review

# Description

Closes #1265 by:

1. **adding a Semgrep rule `untranslated-gui-string`** to catch common cases of GUI strings not marked with gettext's `_()` wrapper; and
2. **adding missing gettext wrappers** to get the Semgrep rule to pass.  :-)

# Test Plan

- [x] The client runs.

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [x] I have tested these changes in the appropriate Qubes environment
 - ~I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)~
 - ~These changes should not need testing in Qubes~

If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:

 - ~I have updated the [AppArmor profile](https://github.com/freedomofpress/securedrop-client/blob/HEAD/files/usr.bin.securedrop-client)~
 - [x] No update to the AppArmor profile is required for these changes
 - ~I don't know and would appreciate guidance~

If these changes modify the database schema, you should include a database migration. Please check as applicable:

 - ~I have written a migration and upgraded a test database based on `main` and confirmed that the migration applies cleanly~
 - ~I have written a migration but have not upgraded a test database based on `main` and would like the reviewer to do so~
 - ~I need help writing a database migration~
 - [x] No database schema changes are needed

### CODE REVIEW

File: securedrop_client/app.py [hunk 1]
```diff
@@ -197,7 +198,7 @@ def start_app(args, qt_args) -> None:
     logging.info("Starting SecureDrop Client {}".format(__version__))
 
     app = QApplication(qt_args)
-    app.setApplicationName("SecureDrop Client")
+    app.setApplicationName(_("SecureDrop Client"))
```
Comment (rmol) [inline securedrop_client/app.py:201]: We have in the past suggested not translating/transliterating product names like "SecureDrop". I think that would hold here.

File: securedrop_client/app.py [hunk 2]
```diff
@@ -169,7 +170,7 @@ def prevent_second_instance(app: QApplication, unique_name: str) -> None:
     except OSError as e:
         if e.errno == ALREADY_BOUND_ERRNO:
             err_dialog = QMessageBox()
-            err_dialog.setText(app.applicationName() + " is already running.")
+            err_dialog.setText(app.applicationName() + _(" is already running."))
```
Comment (rmol) [inline securedrop_client/app.py:173]: Making this something like `setText(_("{application_name} is already running".format(app.applicationName)))` would produce a message that would be easier to translate.

File: .semgrep/custom-rules.yml [hunk 3]
```diff
@@ -97,3 +97,45 @@ rules:
     - pattern-not-inside: |
         def safe_move(...):
             ...
+
+- id: untranslated-gui-string
+  languages:
+    - python
+  severity: ERROR
+  message: Untranslated GUI string
```
Comment (creviera) [inline .semgrep/custom-rules.yml:105]: good idea, thanks for adding this to helps us remember! :)

File: .semgrep/custom-rules.yml [hunk 4]
```diff
@@ -97,3 +97,45 @@ rules:
...
+  patterns:
+
+    # Exempt empty strings from translation:
+    - pattern-not-regex: |
+        \(""\)
```
Comment (creviera) [inline .semgrep/custom-rules.yml:110]: also a cute one-line ascii emoji

**Review by rmol — CHANGES_REQUESTED** [review]
This looks good. We could go down a regex rabbit hole trying to catch cases like `setText(stringvar + "...")` but I think that effort would be better spent on developer guidelines.

I requested a couple of small changes on the very last commit.

File: securedrop_client/gui/widgets.py [hunk 5]
```diff
@@ -747,7 +747,7 @@ def __init__(self):
         bullet1_layout = QHBoxLayout()
         bullet1_layout.setContentsMargins(0, 0, 0, 0)
         bullet1.setLayout(bullet1_layout)
-        bullet1_bullet = QLabel("·")
+        bullet1_bullet = QLabel("·")  # nosemgrep: semgrep.untranslated-gui-string
```
Comment (creviera) [inline securedrop_client/gui/widgets.py:750]: clear example of how this works

**Review by creviera — APPROVED** [review]
lgtm, test plan checks out

File: securedrop_client/app.py (hunk 1 above)
Comment (creviera) [reply securedrop_client/app.py:201]: oh, good catch @rmol!

File: securedrop_client/app.py (hunk 2 above)
Comment (creviera) [reply securedrop_client/app.py:173]: we also have suggested to developers to use f-strings

Comment (rmol) [reply securedrop_client/app.py:173]: that's a good point. the `semgrep` patterns will need to be adjusted to cover those.

Comment (cfm) [reply securedrop_client/app.py]: @rmol, re: https://github.com/freedomofpress/securedrop-client/pull/1272#discussion_r682079513:  Thanks for catching this—amended.

@creviera, re: f-strings:  I'm under the impression that the f-string interpolation will always preempt the `gettext()` call, which is the opposite order of operations we want—at least without some [fancy footwork](https://bugs.python.org/issue36310#msg348417) with a code smell I don't love.  (Cf. also python-babel/babel#594.)  Is this the approach we want to take?

Comment (rmol) [reply securedrop_client/app.py]: @cfm I think you're right, I'm having trouble thinking of a case where f-strings make sense in a translation context. It's more likely they'd just be used to interpolate already-translated strings into another, not be translated themselves, so wouldn't be something we want to catch with semgrep. (And you made me realize my example above would have the same problem -- one of the parentheses at the end should instead close the `_()` call  before `.format`.)

**Comment by cfm** [comment]
`test-buster` is failing on `safety`.

File: securedrop_client/app.py (hunk 2 above)
Comment (rmol) [reply securedrop_client/app.py]: Very nice. I'm happy with this. @creviera ?

**Review by rmol — APPROVED** [review]

**Comment by creviera** [comment]
looks like the only ci failure is due to a know safety issue:

> -> coverage, installed 4.5.1, affected <6.0b1, id 41002

File: securedrop_client/app.py (hunk 2 above)
Comment (creviera) [reply securedrop_client/app.py]: :100: like the checks and the reasoning makes sense

**Pull request merged**